### PR TITLE
import from private registry (Condition work)

### PR DESCRIPTION
**Pull request by goern**
## Related Issues and Dependencies
This PR is implementing conditions for the import use case

## This introduces a breaking change
- No

## This should yield a new module release

- Yes

## This Pull Request implements
partly #103

### CODE REVIEW

**Comment by goern** [comment]
short demo and a little bit of overview at https://www.youtube.com/watch\?v\=phfoEWsl5nk

File: config/samples/meteor_v1alpha1_customnbimage.yaml [hunk 1]
```diff
@@ -72,6 +72,31 @@ spec:
 ---
 apiVersion: meteor.zone/v1alpha1
 kind: CustomNBImage
+metadata:
+  name: s2i-minimal-py38-notebook-import-with-secret # TODO
```
Comment (codificat) [inline config/samples/meteor_v1alpha1_customnbimage.yaml:76]: Like :+1: 

However, I also like having `ImageImport` samples in the separate file you are also adding, so I think this (duplicate) should be removed from here.

Maybe go one step further and separate each sample to its own file to make it easier to pick e.g. for demos? That can be a separate PR though - for now I would say let's not add duplicate content

**Comment by goern** [comment]
some references

https://alenkacz.medium.com/kubernetes-operator-best-practices-implementing-observedgeneration-250728868792

File: api/v1alpha1/cnbi_condition_consts.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
+/*
```
Comment (codificat) [inline api/v1alpha1/cnbi_condition_consts.go:1]: Question: why not leave all the CNBi-related types and consts within `customnbimage_types.go`?

**Review by codificat — COMMENTED** [review]
Many thanks for the work (including the demo)!

So far I only have a couple of doubts about the separation of types/files and the rename of the status field.

Other than that, I am trying to test it but it is not working correctly on my cluster for some reason: the CNBi object does not get udpated conditions/phase:

```
$ oc get pr,cnbi
NAME                                                                  SUCCEEDED   REASON      STARTTIME   COMPLETIONTIME
pipelinerun.tekton.dev/cnbi-s2i-minimal-py38-notebook-import-import   True        Succeeded   2m25s       61s

NAME                                                         PHASE
customnbimage.meteor.zone/s2i-minimal-py38-notebook-import   Pending
```

still looking into it, which might yield some more comments about the actual code.

**Review by codificat — COMMENTED** [review]
Some additional comments

File: api/v1alpha1/cnbi_condition_consts.go (hunk 2 above)
Comment (goern) [reply api/v1alpha1/cnbi_condition_consts.go:1]: for clearity

File: controllers/cnbi/controller.go [hunk 3]
```diff
@@ -303,37 +317,51 @@ func (r *CustomNBImageReconciler) ReconcilePipelineRun(name string, ctx *context
...
 			logger.Error(nil, "Tekton reported multiple conditions")
 		}
-		condition := pipelineRun.Status.Conditions[0]
-		updateStatus(metav1.ConditionStatus(condition.Status), condition.Reason, condition.Message)
+
+		if pipelineRun.Labels["cnbi.thoth-station.ninja/pipeline"] == "import" {
+			if pipelineRun.Status.Conditions[0].Status == v1.ConditionFalse && pipelineRun.Status.Conditions[0].Type == "Succeeded" {
+				logger.Info("Import pipeline failed")
+				setCondition(newStatus, meteorv1alpha1.ImageImportReady, metav1.ConditionFalse, "ImageImportNotReady", "Import failed, this could be due to the repository to import from does not exist or is not accessible")
+				setCondition(newStatus, meteorv1alpha1.PipelineRunCompleted, metav1.ConditionTrue, "PipelineRunCompleted", "The PipelineRun has been completed, Image is importend")
```
[the remark this answers is not in the document]
Comment (goern) [reply controllers/cnbi/controller.go:343]: > however, running a test here with the "import only" sample: I see the the pipeline/cnbi fail, but the ImageStream is created anyway (i.e. the image does get imported).

this is a second comment we should keep separate, as it implies refactoring the import pipeline and rethinking how we handle the imagestream resources

**Review by VannTen — CHANGES_REQUESTED** [review]
Review only on the Makefile.

Besides, those changes (of the Makefile) should IMO be in a separate PR (this one is big enough as is)

File: api/v1alpha1/cnbi_condition_consts.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	// ImageImportReady indicates that the imported image is ready to be used
+	ImageImportReady = ConditionType("ImageImportReady")
+
+	// ImageImportInvalid indicates that the imported image is invalid
+	ImageImportInvalid = ConditionType("ImageImportInvalid")
+
+	// PreparingImageBuild indicates that the image build is being prepared by a Tekton PipelineRun's Step
+	PreparingImageBuild = ConditionType("PreparingImageBuild")
+
+	// ErrorPreparingImageBuild indicates that the image build preparation failed
```
Comment (VannTen) [inline api/v1alpha1/cnbi_condition_consts.go:46]: as above

File: api/v1alpha1/customnbimage_types.go [hunk 5]
```diff
@@ -113,16 +103,21 @@ type CustomNBImageSpec struct {
 	BuildTypeSpec `json:",inline"`
 }
 
-// CustomNBImageStatus defines the observed state of CustomNBImage
-type CustomNBImageStatus struct {
-	// Current condition of the Shower.
+//+kubebuilder:object:generate=true
+// CustomNotebookImageStatus defines the observed state of CustomNBImage
+type CustomNotebookImageStatus struct {
```
[the remark this answers is not in the document]
Comment (VannTen) [reply api/v1alpha1/customnbimage_types.go:108]: Agreed. We should keep NB in that PR, especially since we're gonna change it again in #115 ; this should make the diff a bit more readable

File: api/v1alpha1/customnbimage_types.go [hunk 6]
```diff
@@ -212,3 +224,22 @@ func (r *CustomNBImageRuntimeSpec) isValid() bool {
...
+// hasValidImagePullSecret checks if the ImagePullSecret is valid, eg name is not empty
+func (b *BuildTypeSpec) hasValidImagePullSecret() bool {
+	if b.ImagePullSecret != (ImagePullSecret{}) {
+		if b.ImagePullSecret.Name == "" {
+			return false
+		}
+	}
+
+	return true
+}
```
Comment (VannTen) [inline api/v1alpha1/customnbimage_types.go:237]: Should that not be the job of an admission controller instead ?

File: controllers/cnbi/controller.go [hunk 7]
```diff
@@ -303,37 +317,54 @@ func (r *CustomNBImageReconciler) ReconcilePipelineRun(name string, ctx *context
...
-		updateStatus(metav1.ConditionFalse, "ErrorPipelineRun", fmt.Sprintf("Reconcile resulted in error. %s", err))
-		return err
+		logger.Error(err, "Error fetching PipelineRun")
+		setCondition(newStatus, meteorv1alpha1.GenericPipelineError, metav1.ConditionTrue, "PipelineRunGenericError", err.Error())
+		return newStatus
 	}
 
 	if len(pipelineRun.Status.Conditions) > 0 {
-		if len(pipelineRun.Status.Conditions) != 1 {
+		if len(pipelineRun.Status.Conditions) != 1 { // TODO observe tekton project if they stay with just one condition all the time
```
Comment (VannTen) [inline controllers/cnbi/controller.go:336]: Does that mean: we should add a feature watching for some behaviour of the tekton operator, *or* we should check later if the tekton project change the fact that they set one condition at a time ? I'm not sure from the formulation

File: controllers/cnbi/controller.go [hunk 8]
```diff
@@ -303,37 +317,54 @@ func (r *CustomNBImageReconciler) ReconcilePipelineRun(name string, ctx *context
...
+
+		if pipelineRun.Labels["cnbi.thoth-station.ninja/pipeline"] == "import" {
+			if pipelineRun.Status.Conditions[0].Status == v1.ConditionFalse && pipelineRun.Status.Conditions[0].Type == "Succeeded" {
+				setCondition(newStatus, meteorv1alpha1.ImageImportReady, metav1.ConditionFalse, "ImageImportNotReady", "Import failed, this could be due to the repository to import from does not exist or is not accessible")
+				setCondition(newStatus, meteorv1alpha1.PipelineRunCompleted, metav1.ConditionTrue, "PipelineRunCompleted", "The PipelineRun has been completed, but the Image could not be imported!")
+				removeCondition(newStatus, meteorv1alpha1.PipelineRunCreated)
+			} else if pipelineRun.Status.Conditions[0].Status == v1.ConditionTrue && pipelineRun.Status.Conditions[0].Type == "Succeeded" {
+				setCondition(newStatus, meteorv1alpha1.ImageImportReady, metav1.ConditionTrue, "ImageImportReady", "Import succeeded, the image is ready to be used")
+				setCondition(newStatus, meteorv1alpha1.PipelineRunCompleted, metav1.ConditionTrue, "PipelineRunCompleted", "The PipelineRun has been completed, Image is importend")
+				removeCondition(newStatus, meteorv1alpha1.PipelineRunCreated)
```
Comment (VannTen) [inline controllers/cnbi/controller.go:348]: Should we not check directly the ImageStream to see if our image is ready and available ?

File: controllers/cnbi/controller_test.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+					err := k8sClient.Get(ctx, lookupKey, createdCNBi)
+					return err == nil
+				}, timeout, interval).Should(BeTrue())
+
+				By("looking if the Controller started reconciling the CustomNBImage object")
+				Expect(createdCNBi.Status.Phase).Should(Equal(meteorv1alpha1.CNBiPhaseRunning))
+				Expect(createdCNBi.Status.Conditions).ShouldNot(BeEmpty())
+				Expect(createdCNBi.Status.Conditions[0].Type).Should(Equal(meteorv1alpha1.RequiredSecretMissing))
+			})
+		}) */
```
Comment (VannTen) [inline controllers/cnbi/controller_test.go:154]: If it's commented out, maybe it should not be in this PR but in a future one ? Not fan of including commented code.

File: api/v1alpha1/customnbimage_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (goern) [reply api/v1alpha1/customnbimage_types.go:108]: ACK.

File: Makefile [hunk 10]
```diff
@@ -245,10 +245,15 @@ kind-create: | $(KIND)
 	curl -s https://api.hub.tekton.dev/v1/resource/tekton/task/openshift-client/0.2/raw | sed -e s/Task/ClusterTask/ | kubectl apply -f -
 
 .PHONY: kind-load-img
-kind-load-img: docker-build kind-create
+kind-load-img: docker-build docker-push kind-create
```
[the remark this answers is not in the document]
Comment (goern) [reply Makefile:248]: why not? I'd rather say, we dont need the dependency on `docker-push` here?!

File: api/v1alpha1/customnbimage_types.go (hunk 6 above)
Comment (goern) [reply api/v1alpha1/customnbimage_types.go:237]: It should be checked during reconciliation, as the secret must not exist at CNBi resource creating time. see [#111](https://github.com/thoth-station/meteor-operator/issues/111)

File: api/v1alpha1/customnbimage_types_test.go [hunk 11]
```diff
@@ -108,3 +108,247 @@ func TestIsValid(t *testing.T) {
...
+							Type:   ImageImportReady,
+							Status: metav1.ConditionFalse,
+							Reason: "ImageImportNotReady",
+						},
+					},
+				},
+			},
+			expectedOutput: PhaseRunning,
+		},
+		"importing_missing_secret": {
```
[the remark this answers is not in the document]
Comment (goern) [reply api/v1alpha1/customnbimage_types_test.go:230]: :) it's a human coding...

File: controllers/cnbi/controller.go (hunk 8 above)
Comment (goern) [reply controllers/cnbi/controller.go:348]: let's review if storing the state in the ImageStream is really a good solution.

File: controllers/cnbi/controller.go (hunk 7 above)
Comment (goern) [reply controllers/cnbi/controller.go:336]: #techDebt

**Comment by VannTen** [comment]
I'm getting weird results on the same test we talked about Friday, somehow it fails with a 80s timeout (with the Eventually commit) then succeds in 6s on the next `make test`.
Am I the only one ? I feel a bit crazy.

**Comment by goern** [comment]
> Am I the only one ? I feel a bit crazy.

:( nooooo

**Comment by VannTen** [comment]
Then maybe we have some kind of live/deadlock in the logic somwhere which got the controller stuck an prevent it to update the CNBi depending on timing ?

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (goern) [reply Makefile:248]: argh... ok, re-read your comment, you are right :)

File: Makefile [hunk 12]
```diff
@@ -245,10 +245,15 @@ kind-create: | $(KIND)
...
 	@echo "Loading image into kind"
 	$(KIND) load docker-image $(IMG) --name $(KIND_CLUSTER_NAME)
 
+.PHONY: kind-copy-token
+kind-copy-token: kind
```
[the remark this answers is not in the document]
Comment (VannTen) [reply Makefile:253]: Besides, do we really need a Makefile target for that ?

**Comment by VannTen** [comment]
Relevant output of the failing test:
```
CustomNBImage controller
/home/max/redhat/thoth-station/meteor-operator/controllers/cnbi/controller_test.go:37
  when a CustomNBImage object is created with a RuntimeEnvironment and a PackageList
  /home/max/redhat/thoth-station/meteor-operator/controllers/cnbi/controller_test.go:44
    [It] should be in Phase 'Running'
    /home/max/redhat/thoth-station/meteor-operator/controllers/cnbi/controller_test.go:47

  Begin Captured GinkgoWriter Output >>
    STEP: creating a CustomNBImage object 11/08/22 10:44:41.13
    1.66790068113038e+09	INFO	Starting EventSource	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "source": "kind source: *v1alpha1.CustomNBImage"}
    1.6679006811304061e+09	INFO	Starting EventSource	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "source": "kind source: *v1beta1.PipelineRun"}
    1.6679006811304097e+09	INFO	Starting EventSource	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "source": "kind source: *v1alpha1.Meteor"}
    1.6679006811304116e+09	INFO	Starting Controller	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage"}
    1.6679006812314403e+09	INFO	Starting workers	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "worker count": 1}
    1.667900683164848e+09	INFO	Creating PipelineRun	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "customNBImage": {"name":"test-1","namespace":"default"}, "namespace": "default", "name": "test-1", "reconcileID": "2f4aa878-0d8b-4235-adac-d31c12356c5d", "pipelinerun": "default/cnbi-test-1-package-list"}
    1.667900683170949e+09	INFO	Reconciled CustomNotebookImage	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "customNBImage": {"name":"test-1","namespace":"default"}, "namespace": "default", "name": "test-1", "reconcileID": "2f4aa878-0d8b-4235-adac-d31c12356c5d", "spec": {"runtimeEnvironment":{"pythonVersion":"3.8","osName":"ubi","osVersion":"8"},"packageVersions":["numpy","pandas","scikit-learn"],"buildType":"PackageList","imagePullSecret":{"name":""}}, "status": {"phase":"Pending","pipelines":[{"name":"package-list","pipelineRunName":"cnbi-test-1-package-list","ready":"False"}]}}
    1.667900683170968e+09	INFO	updating status of CustomNotebookImage	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "customNBImage": {"name":"test-1","namespace":"default"}, "namespace": "default", "name": "test-1", "reconcileID": "2f4aa878-0d8b-4235-adac-d31c12356c5d", "status": {"phase":"Pending","pipelines":[{"name":"package-list","pipelineRunName":"cnbi-test-1-package-list","ready":"False"}]}, "newStatus": {"observedGeneration":1,"phase":"Pending","conditions":[{"type":"PipelineRunCreated","status":"True","lastTransitionTime":"2022-11-08T09:44:43Z","reason":"PipelineRunCreated","message":"package-list PipelineRun created successfully"}]}}
    1.6679006831755385e+09	INFO	Reconciled CustomNotebookImage	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "customNBImage": {"name":"test-1","namespace":"default"}, "namespace": "default", "name": "test-1", "reconcileID": "539671d4-a57e-45e1-b533-41a25e2f1b30", "spec": {"runtimeEnvironment":{"pythonVersion":"3.8","osName":"ubi","osVersion":"8"},"packageVersions":["numpy","pandas","scikit-learn"],"buildType":"PackageList","imagePullSecret":{"name":""}}, "status": {"phase":"Pending","pipelines":[{"name":"package-list","pipelineRunName":"cnbi-test-1-package-list","ready":"False"}]}}
    1.6679006831755712e+09	INFO	updating status of CustomNotebookImage	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "customNBImage": {"name":"test-1","namespace":"default"}, "namespace": "default", "name": "test-1", "reconcileID": "539671d4-a57e-45e1-b533-41a25e2f1b30", "status": {"phase":"Pending","pipelines":[{"name":"package-list","pipelineRunName":"cnbi-test-1-package-list","ready":"False"}]}, "newStatus": {"observedGeneration":1,"phase":"Pending"}}
    1.6679006831910357e+09	INFO	Reconciled CustomNotebookImage	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "customNBImage": {"name":"test-1","namespace":"default"}, "namespace": "default", "name": "test-1", "reconcileID": "92560079-933e-42d8-ada4-bfaa275a5738", "spec": {"runtimeEnvironment":{"pythonVersion":"3.8","osName":"ubi","osVersion":"8"},"packageVersions":["numpy","pandas","scikit-learn"],"buildType":"PackageList","imagePullSecret":{"name":""}}, "status": {"observedGeneration":1,"phase":"Pending","pipelines":[{"name":"package-list","pipelineRunName":"cnbi-test-1-package-list","ready":"False"}]}}
    1.6679006831910474e+09	INFO	updating status of CustomNotebookImage	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "customNBImage": {"name":"test-1","namespace":"default"}, "namespace": "default", "name": "test-1", "reconcileID": "92560079-933e-42d8-ada4-bfaa275a5738", "status": {"observedGeneration":1,"phase":"Pending","pipelines":[{"name":"package-list","pipelineRunName":"cnbi-test-1-package-list","ready":"False"}]}, "newStatus": {"observedGeneration":1,"phase":"Pending"}}
    1.6679006831924095e+09	INFO	Reconciled CustomNotebookImage	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "customNBImage": {"name":"test-1","namespace":"default"}, "namespace": "default", "name": "test-1", "reconcileID": "95ee7c23-e684-49fb-9fa8-03c8207bb78e", "spec": {"runtimeEnvironment":{"pythonVersion":"3.8","osName":"ubi","osVersion":"8"},"packageVersions":["numpy","pandas","scikit-learn"],"buildType":"PackageList","imagePullSecret":{"name":""}}, "status": {"observedGeneration":1,"phase":"Pending","pipelines":[{"name":"package-list","pipelineRunName":"cnbi-test-1-package-list","ready":"False"}]}}
    1.66790068319242e+09	INFO	updating status of CustomNotebookImage	{"controller": "customnbimage", "controllerGroup": "meteor.zone", "controllerKind": "CustomNBImage", "customNBImage": {"name":"test-1","namespace":"default"}, "namespace": "default", "name": "test-1", "reconcileID": "95ee7c23-e684-49fb-9fa8-03c8207bb78e", "status": {"observedGeneration":1,"phase":"Pending","pipelines":[{"name":"package-list","pipelineRunName":"cnbi-test-1-package-list","ready":"False"}]}, "newStatus": {"observedGeneration":1,"phase":"Pending"}}
  << End Captured GinkgoWriter Output

  Timed out after 80.000s.
  Expected success, but got an error:
      <*errors.errorString | 0xc000abb760>: {
          s: "Assertion in callback at /home/max/redhat/thoth-station/meteor-operator/controllers/cnbi/controller_test.go:70 failed:\nExpected\n    <[]v1alpha1.Condition | len:0, cap:0>: nil\nnot to be empty",
      }
      Assertion in callback at /home/max/redhat/thoth-station/meteor-operator/controllers/cnbi/controller_test.go:70 failed:
      Expected
          <[]v1alpha1.Condition | len:0, cap:0>: nil
      not to be empty
  In [It] at: /home/max/redhat/thoth-station/meteor-operator/controllers/cnbi/controller_test.go:72
------------------------------
```

**Comment by VannTen** [comment]
By the way, using the following patch make the test fail more consistently (1/2) but still not all the time.
```diff
diff --git a/controllers/cnbi/controller_test.go b/controllers/cnbi/controller_test.go
index f6e5240..ab4b601 100644
--- a/controllers/cnbi/controller_test.go
+++ b/controllers/cnbi/controller_test.go
@@ -30,7 +30,7 @@ import (
 )
 
 const (
-	timeout  = time.Second * 80
+	timeout  = time.Second * 30
 	interval = time.Millisecond * 750
 )
 
@@ -63,12 +63,14 @@ var _ = Describe("CustomNBImage controller", func() {
 
 			lookupKey := types.NamespacedName{Name: "test-1", Namespace: "default"}
 
-			Eventually(func(g Gomega) {
-				createdCNBi := &meteorv1alpha1.CustomNBImage{}
-				err := k8sClient.Get(ctx, lookupKey, createdCNBi)
-				g.Expect(err).NotTo(HaveOccurred())
-				g.Expect(createdCNBi.Status.Conditions).ToNot(BeEmpty())
-				g.Expect(createdCNBi.Status.Phase).To(Equal(meteorv1alpha1.PhaseRunning))
+			Eventually(func(gg Gomega) {
+				gg.Consistently(func(g Gomega) {
+					createdCNBi := &meteorv1alpha1.CustomNBImage{}
+					err := k8sClient.Get(ctx, lookupKey, createdCNBi)
+					g.Expect(err).NotTo(HaveOccurred())
+					g.Expect(createdCNBi.Status.Conditions).ToNot(BeEmpty())
+					g.Expect(createdCNBi.Status.Phase).To(Equal(meteorv1alpha1.PhaseRunning))
+				}, "8s", "500ms").Should(Succeed())
 			}, timeout, interval).Should(Succeed())
 
 		})
```

(This basically check taht the our conditions stay true)
("Eventual consistency" takes its whole sense here ^^)

File: controllers/cnbi/controller.go [hunk 13]
```diff
@@ -112,42 +110,36 @@ func (r *CustomNBImageReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		}
+		original.Status = *status
+		if err := r.Client.Status().Update(ctx, original); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("failed to update status of CNBi %s/%s: %v", nn.Namespace, nn.Name, err)
+	}
+	return nil
```
Comment (VannTen) [inline controllers/cnbi/controller.go:130]: I wonder if this might not be the root of (or at least contributing to) the failing test.
The doc for RetryOnConflict says :
```
// RetryOnConflict is used to make an update to a resource when you have to worry about
// conflicts caused by other code making unrelated updates to the resource at the same
// time. fn should fetch the resource to be modified, make appropriate changes to it, try
// to update it, and return (unmodified) the error from the update function. On a
```
But here, we get the resource at the start of the reconcile loop, mutate stuff (conditions etc) then re-get the resource and try to modify. Should we instead wrap the whole reconcile function into RetryOnConflict ?

Comment (goern) [reply controllers/cnbi/controller.go:130]: ja, if we return an err from Reconcile() it will be requeued automatically. If any of the owned resources changes, this controller will reconcile, see line 99

Comment (goern) [reply controllers/cnbi/controller.go:130]: ... I am still thinking if line 92 is correct...

Comment (VannTen) [reply controllers/cnbi/controller.go:130]: > ja, if we return an err from Reconcile() it will be requeued automatically. If **any of the owned resources changes**, this controller will reconcile, see line 99

By owned resources you mean PipelineRun with `ownerRef` ? Line 99 will only requeue reconciliation if `err != nil` AFAICT.

**Comment by VannTen** [comment]
It seems the sections `pipelines` and `conditions` of the CNBi status are not updated the same way.
Pipelines are update in controller.go#L155, (inside the reconcilePipelineRun), and  conditions are in a `newStatus` which replace the previous one (line 123)
But as far as I can tell from the code the pipelines part of the status is never sent to the API server, we just the state of the internal struct in the controller.

I still don't get why the conditions do not stick though... I tried to replace the newstatus with just using the internal CNBi field (the one defined in CustomNBImageReconciler) and using that for updating the API instead, and it does seem to works and get rid of the error, but I'm still not sure of the root cause...

**Comment by codificat** [comment]
I still get flaky test results, failing sometimes. But a few functionality tests all have correctly propagated pipeline status to cnbi phase, so I'd be ok merging this to move forward and address the flaky tests later.

/approve

File: api/v1alpha1/common.go [hunk 14]
```diff
@@ -16,10 +16,15 @@ limitations under the License.
...
-	PhaseFailed    = "Failed"
-	PhaseResolving = "Resolving"
-	PhaseBuilding  = "Building"
-	PhaseOk        = "Ready"
+	PhasePending   = Phase("Pending")
+	PhaseFailed    = Phase("Failed")
+	PhaseRunning   = Phase("Running")
+	PhaseBuilding  = Phase("Building")
+	PhaseSucceeded = Phase("Succeeded")
+	PhaseUnknown   = Phase("Unknown")
```
Comment (VannTen) [inline api/v1alpha1/common.go:29]: From the same document (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) (search for `phase`) it looks like the Phase pattern is deprecated and unadvised. This is the advice for core Kubernetes types (I think), but the motivation probably apply to us as well...

> Some resources in the v1 API contain fields called phase, and associated message, reason, and other status fields. The pattern of using phase is deprecated. Newer API types should use conditions instead. Phase was essentially a state-machine enumeration field, that contradicted [system-design principles](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/principles.md#control-logic) and hampered evolution, since [adding new enum values breaks backward compatibility](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md). Rather than encouraging clients to infer implicit properties from phases, we prefer to explicitly expose the individual conditions that clients need to monitor. Conditions also have the benefit that it is possible to create some conditions with uniform meaning across all resource types, while still exposing others that are unique to specific resource types.

**Comment by VannTen** [comment]
>  I'd be ok merging this to move forward and address the flaky tests later.

Same. It's not like we're gonna break prod ^. Additionnaly, Github handling of dependent PR is non-existent, so working on this PR as a base is not super practical.

File: api/v1alpha1/cnbi_condition_types.go [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+	// Last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,3,opt,name=lastTransitionTime"`
+	// Unique, one-word, CamelCase reason for the condition's last transition.
+	// +optional
+	Reason string `json:"reason,omitempty" protobuf:"bytes,4,opt,name=reason"`
+	// Human-readable message indicating details about last transition.
+	// +optional
+	Message string `json:"message,omitempty" protobuf:"bytes,5,opt,name=message"`
+}
```
[the remark this answers is not in the document]
Comment (goern) [reply api/v1alpha1/cnbi_condition_types.go:46]: we could, I was following the pattern of typed Conditions, so that we have ConditionType (which is a string).

File: api/v1alpha1/common.go (hunk 14 above)
Comment (goern) [reply api/v1alpha1/common.go:29]: ACK, the Phase pattern is kept because it give and easy overview of the state (via AggregatePhase()), I would also consider it deprecated and remove it later on...

**Comment by VannTen** [comment]
/lgtm

As discussed, we can address the issues we have in future PRs

**Comment by goern** [comment]
looks like thoth-station/prow/make-test is failing due to prow problems?!

**Comment by codificat** [comment]
The build-check is failing, and https://github.com/thoth-station/meteor-operator/pull/125 should resolve that.

The make-test I believe is not blocking the PR merge... I'll look into it now but the above PR should hopefully let us merge this one

**Comment by codificat** [comment]
> looks like thoth-station/prow/make-test is failing due to prow problems?!

:shrug: it might be... the last make-test prow jobs failed with:

```
 Failed to open /logs/process-log.txt: open /logs/process-log.txt: no such file or directory 
```

but the make-test job in #125 just completed fine. It might have been a temporary glitch.

The issue with build-check is real though, so let's get #125 in and then this one can hopefully pass all checks and get merged

**Comment by codificat** [comment]
Meh

Build-check failed again with the same error:

```
+ set +e
++ buildah bud --tls-verify=false --layers -f /gen-source/Containerfile -t meteor-operator-110 .
+ out='[1/2] STEP 1/10: FROM golang:1.18 AS builder
error creating build container: short-name resolution enforced but cannot prompt without a TTY'
+ exit_code=125
+ set -e
+ [[ 125 -ne 0 ]]
+ state=failure
+ desc='The image build test failed!'
+ cat
```

a rebase of the PR should avoid that (maybe an issue could be raised for aicoe-ci to work without a rebase, too).

make-test failed too: this time the test actually ran - but we hit the flaky test:

```
  Timed out after 80.001s.
  Expected success, but got an error:
      <*errors.errorString | 0xc000d686b0>: {
          s: "Assertion in callback at /home/prow/go/src/github.com/thoth-station/meteor-operator/controllers/cnbi/controller_test.go:70 failed:\nExpected\n    <[]v1alpha1.Condition | len:0, cap:0>: nil\nnot to be empty",
      }
      Assertion in callback at /home/prow/go/src/github.com/thoth-station/meteor-operator/controllers/cnbi/controller_test.go:70 failed:
      Expected
          <[]v1alpha1.Condition | len:0, cap:0>: nil
      not to be empty
  In [It] at: /home/prow/go/src/github.com/thoth-station/meteor-operator/controllers/cnbi/controller_test.go:72
```

Next steps then:
1. rebase PR to include latest Dockerfile changes with full image ref
2. `/retest`
3. ... *hope not to hit the flaky test* :crossed_fingers:
4. Profit!

**Comment by VannTen** [comment]
Can't we just manually merge it and bypass the flaky test ?
